rust/parquet/src/arrow/levels.rs

```diff
@@ -294,40 +242,1158 @@ impl LevelInfo {
+            vec![Arc::new(a), Arc::new(b), Arc::new(c)],
+        )
+        .unwrap();
+
+        //////////////////////////////////////////////
+        let expected_batch_level = LevelInfo {
+            definition: vec![0; 5],
+            repetition: None,
+            array_offsets: (0..=5).collect(),
+            array_mask: vec![true, true, true, true, true],
+            max_definition: 0,
+            is_list: false,
+            is_nullable: false,
+        };
+
+        let batch_level = LevelInfo::new_from_batch(&batch);
+        assert_eq!(&batch_level, &expected_batch_level);
+
+        // calculate the list's level
+        let mut levels = vec![];
+        batch
+            .columns()
+            .iter()
+            .zip(batch.schema().fields())
+            .for_each(|(array, field)| {
+                let mut array_levels = batch_level.calculate_array_levels(array, field);
+                levels.append(&mut array_levels);
+            });
+        assert_eq!(levels.len(), 5);
+
+        // test "a" levels
+        let list_level = levels.get(0).unwrap();
+
+        let expected_level = LevelInfo {
+            definition: vec![1, 1, 1, 1, 1],
+            repetition: None,
+            array_offsets: vec![0, 1, 2, 3, 4, 5],
+            array_mask: vec![true, true, true, true, true],
+            max_definition: 1,
+            is_list: false,
+            is_nullable: false,
+        };
+        assert_eq!(list_level, &expected_level);
+
+        // test "b" levels
+        let list_level = levels.get(1).unwrap();
+
+        let expected_level = LevelInfo {
+            definition: vec![1, 0, 0, 1, 1],
+            repetition: None,
+            array_offsets: vec![0, 1, 2, 3, 4, 5],
+            array_mask: vec![true, false, false, true, true],
+            max_definition: 1,
+            is_list: false,
+            is_nullable: true,
+        };
+        assert_eq!(list_level, &expected_level);
+
+        // test "d" levels
+        let list_level = levels.get(2).unwrap();
+
+        let expected_level = LevelInfo {
+            definition: vec![1, 1, 1, 2, 1],
+            repetition: None,
+            array_offsets: vec![0, 1, 2, 3, 4, 5],
+            array_mask: vec![false, false, false, true, false],
+            max_definition: 2,
+            is_list: false,
+            is_nullable: true,
+        };
+        assert_eq!(list_level, &expected_level);
+
+        // test "f" levels
+        let list_level = levels.get(3).unwrap();
+
+        let expected_level = LevelInfo {
+            definition: vec![3, 2, 3, 2, 3],
+            repetition: None,
+            array_offsets: vec![0, 1, 2, 3, 4, 5],
+            array_mask: vec![true, false, true, false, true],
+            max_definition: 3,
+            is_list: false,
+            is_nullable: true,
+        };
+        assert_eq!(list_level, &expected_level);
+    }
+
+    #[test]
+    fn test_filter_array_indices() {
+        let level = LevelInfo {
```

Comment:
I didn't have enough test cases to compute here. I'll add more when I increase coverage of deeply-nested lists